### ISSUE TITLE
Unable to run rdkafka with Alpine (node:6 and node:8)

### ISSUE THREAD

**Issue by IdanAdar** [gliderlabs/docker-alpine#385]
> Also cross-posted in the rdkafka GitHub repository: https://github.com/Blizzard/node-rdkafka/issues/361

We are using the following as a base image: 
```
FROM node:8-alpine

RUN apk --no-cache add \
      bash \
      g++ \
      ca-certificates \
      lz4-dev \
      musl-dev \
      cyrus-sasl-dev \
      openssl-dev \
      make \
      python
```

We then use this base image in the following Dockerfile:
```
FROM the-base-image

# Create app directory
RUN mkdir -p /usr/src/app
RUN mkdir temp

# Copy package.json first to check if an npm install is needed
COPY .npmrc /temp
COPY package.json /temp
RUN cd temp && npm install --production --unsafe-perm
RUN cp -a /temp/node_modules /usr/src/app

# Bundle app source
COPY . /usr/src/app

WORKDIR /usr/src/app

ENV PORT 8080
EXPOSE 8080

RUN chmod +x environment.sh
CMD ["/bin/sh", "/usr/src/app/environment.sh"]
```

When the built image is then attempting to start, we get the following error in the log:
```
"message":"kafkaProducerConfig is: '{\"security.protocol\":\"sasl_ssl\",\"sasl.mechanisms\":\"PLAIN\",\"api.version.request\":true,\"broker.version.fallback\":\"0.10.2.1\",\"dr_cb\":false,\"message.send.max.retries\":2,\"retry.backoff.ms\":100,\"socket.keepalive.enable\":true,\"metadata.broker.list\":\"kafka01-prod02.****.****.****.****.net:9093,kafka02-prod02.****.****.****.****.net:9093,kafka03-prod02.****.****.****.****.net:9093,kafka04-prod02.****.****.****.****.net:9093,kafka05-prod02.****.****.****.****.net:9093\",\"sasl.username\":\"*****\",\"sasl.password\":\"****\",\"ssl.ca.location\":\"/etc/ssl/certs\"}'\nkafkaTopicConfig is: '{\"request.required.acks\":0,\"request.timeout.ms\":5000,\"message.timeout.ms\":300000}'\nproducerOpts is: '{\"pollInterval\":100,\"DEFAULTS\":{\"POLL_INTERVAL\":100}}'","instanceId":"SystemId","module":"metadata-producer [Producer]","execTime_date":"2018-02-25T09:25:39.613Z","logLevel":"DEBUG"}
```
```
{"message":"Init error: '\"Couldn't connect to Kafka. Reason is: Error: Failed to create thread: No such file or directory (2). .\"'","instanceId":"SystemId","module":"metadata-producer","execTime_date":"2018-02-25T09:25:39.629Z","logLevel":"INFO"}
```

Has anyone encountered this issue with rdkafka and Alpine and knows of a workaround or solution?

**Comment by IdanAdar** [comment]
Thanks @kaniini, could you please elaborate? How do I implement your suggestion?

**Comment by beanz** [comment]
Try changing:

    CMD ["/bin/sh", "/usr/src/app/environment.sh"]

to:

    RUN apk --no-cache add strace
    CMD ["strace", "-e", "file", "/bin/sh", "/usr/src/app/environment.sh"]

This should produce the strace logs of file accesses to the container stderr.

**Comment by hindessm** [comment]
Try changing:

    CMD ["/bin/sh", "/usr/src/app/environment.sh"]

to:

    RUN apk --no-cache add strace
    CMD ["strace", "-e", "file", "/bin/sh", "/usr/src/app/environment.sh"]

This should produce the strace logs of file accesses to the container stderr.

**Comment by kaniini** [comment]
needs to be "-f", "-e" actually, as you need to follow forks.

**Comment by IdanAdar** [comment]
We're getting the following:

```
[pid    26] open("/usr/lib/sasl2", O_RDONLY|O_DIRECTORY|O_CLOEXEC) = 10
[pid    26] open("/usr/lib/sasl2/libsasldb.la", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid    26] open("/usr/lib/sasl2/libsasldb.so", O_RDONLY|O_CLOEXEC) = 11
[pid    26] open("/lib/libdb-5.3.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid    26] open("/usr/local/lib/libdb-5.3.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid    26] open("/usr/lib/libdb-5.3.so", O_RDONLY|O_CLOEXEC) = 11
[pid    26] open("/usr/lib/sasl2/libplain.la", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid    26] open("/usr/lib/sasl2/libplain.so", O_RDONLY|O_CLOEXEC) = 11
[pid    26] open("/usr/lib/sasl2/liblogin.la", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid    26] open("/usr/lib/sasl2/liblogin.so", O_RDONLY|O_CLOEXEC) = 11
[pid    26] stat("/etc/ssl", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
strace: Process 30 attached
alerts producer is running on port 3001
{"message":"Init error: '\"Couldn't connect to Kafka. Reason is: Error: Failed to create thread: No such file or directory (2). .\"'","instanceId":"SystemId","module":"metadata-producer","execTime_date":"2018-03-04T09:57:08.298Z","logLevel":"INFO"}
```

Are we missing packages?
```
RUN apk --no-cache add \
      bash \
      g++ \
      ca-certificates \
      lz4-dev \
      musl-dev \
      cyrus-sasl-dev \
      openssl-dev \
      make \
      python
```

**Comment by IdanAdar** [comment]
We ended up changing the base Dockerfile to the following, [based on this question](https://github.com/Blizzard/node-rdkafka/issues/111#issuecomment-297236063). Essentially, instead of building librdkafka during the `npm install` stage, we fetch and compile it beforehand. And it seems to be working. 

If we remove or change the marked (***) entries, it fails. We'd love to understand why it then fails as I rather not use "untrusted repositories" and have _some_ understanding of the packages used. 

```
FROM node:8-alpine
RUN apk add --upgrade --no-cache \
  *** yarn --repository http://dl-3.alpinelinux.org/alpine/edge/community/ --allow-untrusted \
  *** libc6-compat \ 
  bash \
  make \
  gcc \
  g++ \
  python \
  cyrus-sasl-dev \
  *** libressl2.5-libcrypto --repository http://dl-3.alpinelinux.org/alpine/edge/main/ --allow-untrusted \
  *** libressl2.5-libssl --repository http://dl-3.alpinelinux.org/alpine/edge/main/ --allow-untrusted \
  *** librdkafka-dev --repository http://dl-3.alpinelinux.org/alpine/edge/community/ --allow-untrusted
  *** dumb-init --repository http://dl-3.alpinelinux.org/alpine/edge/community/ --allow-untrusted 
  ...
  ...
  ENV BUILD_LIBRDKAFKA=0
```

**Comment by IdanAdar** [comment]
Fixed by node-rdkafka in v2.3.2.
See here: https://github.com/Blizzard/node-rdkafka/issues/361
